torch_np/_detail/_reductions.py

```diff
@@ -4,13 +4,78 @@
 Anything here only deals with torch objects, e.g. "dtype" is a torch.dtype instance etc
 """
 
+import typing
+
 import torch
 
 from . import _dtypes_impl, _util
 
 NoValue = None
 
 
+import functools
+
+############# XXX
+### From _util.axis_expand_func
+
+
+def deco_axis_expand(func):
+    """Generically handle axis arguments in reductions."""
+
+    @functools.wraps(func)
+    def wrapped(tensor, axis, *args, **kwds):
+
+        if axis is not None:
+            if not isinstance(axis, (list, tuple)):
+                if not isinstance(axis, typing.SupportsIndex):
+                    raise TypeError(
+                        f"{type(axis)=}, but should be a list/tuple or support operator.index()"
+                    )
+                axis = (axis,)
+            axis = _util.normalize_axis_tuple(axis, tensor.ndim)
+
+        if axis == ():
+            newshape = _util.expand_shape(tensor.shape, axis=0)
+            tensor = tensor.reshape(newshape)
+            axis = (0,)
+
+        result = func(tensor, axis=axis, *args, **kwds)
+        return result
+
+    return wrapped
+
+
+def emulate_keepdims(func):
```

Comment:
This should probably be on the main normalization suit. Then, I reckon it should be alright to simply inline `apply_keepdims` there.

Reply:
`keepdims` interacts with `axis`, so moving this to the main normalization dance would break an invariant that normalization of an argument only depends on this argument and nothing else.

Reply:
I get the feeling that it should be possible to simplify these further. Let's leave as it is for now though. I may put some more thought on this at a later time.